### PR TITLE
sec(curveTemplates): secure page

### PR DESCRIPTION
**Pull request by adr-mo**
## Description

Secure graph template pages (other PR for 2.8.x)

**Fixes** # (issue)

## Type of change

- [x] Patch fixing an issue (non-breaking change)
- [ ] New functionality (non-breaking change)
- [ ] Breaking change (patch or feature) that might cause side effects breaking part of the Software
- [ ] Updating documentation (missing information, typo...)

## Target serie

- [x] 20.04.x
- [x] 20.10.x
- [x] 21.04.x
- [x] 21.10.x (master)

<h2> How this pull request can be tested ? </h2>

How to test is described in the Jira ticket

## Checklist

- [x] I have followed the **coding style guidelines** provided by Centreon
- [x] I have commented my code, especially new **classes**, **functions** or any **legacy code** modified. (***docblock***)
- [ ] I have commented my code, especially **hard-to-understand areas** of the PR.
- [ ] I have made corresponding changes to the **documentation**.
- [x] I have **rebased** my development branch on the base branch (master, maintenance).

### CODE REVIEW

File: www/include/views/componentTemplates/DB-Func.php [hunk 1]
```diff
@@ -150,245 +162,219 @@ function multipleComponentTemplateInDB($compos = array(), $nbrDup = array())
...
+                    } else {
+                        $bindParams[':' . $inputName] = [\PDO::PARAM_STR => $inputValue];
+                    }
+                }
+                break;
+            case 'ds_color_line_mode':
+                $bindParams[':' . $inputName] = [
+                    \PDO::PARAM_STR => in_array($inputValue[$inputName], ['0', '1'])
+                        ? '0'
+                        : $inputValue[$inputName]
```
Comment (kduret) [inline www/include/views/componentTemplates/DB-Func.php:327]:
```suggestion
                        ? $inputValue[$inputName]
                        : '0'
```

File: www/include/views/componentTemplates/DB-Func.php [hunk 2]
```diff
@@ -39,54 +39,44 @@
 
 function NameHsrTestExistence($name = null)
 {
-    global $pearDB, $pearDBO, $form;
-    $gsvs = null;
+    global $pearDB, $form;
+    $formValues = null;
```
Comment (kduret) [inline www/include/views/componentTemplates/DB-Func.php:43]:
```suggestion
    $formValues = [];
```

File: www/include/views/componentTemplates/DB-Func.php [hunk 3]
```diff
@@ -150,245 +162,219 @@ function multipleComponentTemplateInDB($compos = array(), $nbrDup = array())
...
+                    \PDO::PARAM_INT => (filter_var($inputValue, FILTER_VALIDATE_INT) === false)
+                        ? null
+                        : (int) $inputValue
+                ];
+                break;
+            case 'default_tpl1':
+                $bindParams[':' . $inputName] = [
+                    \PDO::PARAM_INT => (filter_var($inputValue, FILTER_VALIDATE_INT) === false)
+                        ? null
+                        : $inputValue
```
Comment (kduret) [inline www/include/views/componentTemplates/DB-Func.php:373]:
```suggestion
                        : (int) $inputValue
```

File: www/include/views/componentTemplates/DB-Func.php [hunk 4]
```diff
@@ -39,54 +39,44 @@
...
+
+    $stmt = $pearDB->prepare($query);
+
+    $stmt->bindValue(':name', $name, \PDO::PARAM_STR);
+
+    foreach ($bindParams as $token => $bindValues) {
+        foreach ($bindValues as $paramType => $value) {
+            $stmt->bindValue($token, $value, $paramType);
+        }
     }
```
Comment (kduret) [inline www/include/views/componentTemplates/DB-Func.php:74]:
```suggestion
    foreach ($bindParams as $token => list($paramType, $value)) {
        $stmt->bindValue($token, $value, $paramType);
    }
```

File: www/include/views/componentTemplates/DB-Func.php [hunk 5]
```diff
@@ -150,245 +162,219 @@ function multipleComponentTemplateInDB($compos = array(), $nbrDup = array())
...
+    $query = 'INSERT INTO `giv_components_template` (`compo_id`, ';
+    $query .= implode(', ', $params) . ') ';
+
+    $query .= 'VALUES (NULL, ' . implode(', ', array_keys($bindParams)) . ')';
+    $stmt = $pearDB->prepare($query);
+    foreach ($bindParams as $token => $bindValues) {
+        foreach ($bindValues as $paramType => $value) {
+            $stmt->bindValue($token, $value, $paramType);
+        }
+    }
```
Comment (kduret) [inline www/include/views/componentTemplates/DB-Func.php:220]:
```suggestion
    foreach ($bindParams as $token => list($paramType, $value)) {
        $stmt->bindValue($token, $value, $paramType);
    }
```

File: www/include/views/componentTemplates/DB-Func.php [hunk 6]
```diff
@@ -150,245 +162,219 @@ function multipleComponentTemplateInDB($compos = array(), $nbrDup = array())
...
-        $ret["ds_color_line_mode"]["ds_color_line_mode"] = '1';
+    $query = rtrim($query, ', ');
+    $query .= ' WHERE compo_id = :compo_id';
+
+    $stmt = $pearDB->prepare($query);
+    foreach ($bindParams as $token => $bindValues) {
+        foreach ($bindValues as $paramType => $value) {
+            $stmt->bindValue($token, $value, $paramType);
+        }
     }
```
Comment (kduret) [inline www/include/views/componentTemplates/DB-Func.php:286]:
```suggestion
    foreach ($bindParams as $token => list($paramType, $value)) {
        $stmt->bindValue($token, $value, $paramType);
    }
```

File: www/include/views/componentTemplates/DB-Func.php [hunk 7]
```diff
@@ -39,54 +39,42 @@
...
 
-function DsHsrTestExistence($name = null)
-{
-    global $pearDB, $pearDBO, $form;
-    $gsvs = null;
-    if (isset($form)) {
-        $gsvs = $form->getSubmitValues();
+    $query .= (empty($formValues['host_id']) && empty($formValues['service_id']))
+        ? ' AND host_id IS NULL  AND service_id IS NULL'
+        : ' AND host_id = :host_id AND service_id = :service_id';
```
Comment (jeremyjaouen) [inline www/include/views/componentTemplates/DB-Func.php:60]: If one of those 2 variables is empty you will go to your else `' AND host_id = :host_id AND service_id = :service_id'` and possibly bind an empty value

File: www/include/views/componentTemplates/DB-Func.php [hunk 8]
```diff
@@ -39,54 +39,42 @@
...
+    foreach ($bindParams as $token => list($paramType, $value)) {
+            $stmt->bindValue($token, $value, $paramType);
     }
-    $dbResult = $pearDB->query($sql);
-    $compo = $dbResult->fetch();
-    #Modif case
-    if ($dbResult->rowCount() >= 1 && $compo["compo_id"] == $gsvs["compo_id"]) {
+    $stmt->execute();
+    $compo = $stmt->fetch();
+    if ($stmt->rowCount() >= 1 && $compo['compo_id'] == $formValues['compo_id']) {
```
Comment (jeremyjaouen) [inline www/include/views/componentTemplates/DB-Func.php:75]: double equals are needed ? If it's possible could you replace it by triple equals?

File: www/include/views/componentTemplates/DB-Func.php [hunk 9]
```diff
@@ -39,54 +39,42 @@
...
         return true;
-    } #Duplicate entry
-    elseif ($dbResult->rowCount() >= 1 && $compo["compo_id"] != $gsvs["compo_id"]) {
+    } elseif ($stmt->rowCount() >= 1 && $compo['compo_id'] != $formValues['compo_id']) {
```
Comment (jeremyjaouen) [inline www/include/views/componentTemplates/DB-Func.php:77]: double equals are needed ? If it's possible could you replace it by triple equals?

**Review by jeremyjaouen — CHANGES_REQUESTED** [review]

**Review by jeremyjaouen — APPROVED** [review]

**Review by kduret — APPROVED** [review]

**Pull request merged**